File: 2025/01/solution.py
def solve_safe_combination(filename):
    """
    Solve the safe combination puzzle (Part 1).

    The dial has positions 0-99 in a circle.
    Start at position 50.
    Process rotations (L=left/lower, R=right/higher).
    Count how many times the dial points at 0 after any rotation.
    """
    with open(filename, 'r') as f:
        rotations = f.read().strip().split('\n')

    position = 50
    zero_count = 0

    for rotation in rotations:
        direction = rotation[0]
        distance = int(rotation[1:])

        if direction == 'L':
            position = (position - distance) % 100
        else:  # direction == 'R'
            position = (position + distance) % 100

        if position == 0:
            zero_count += 1

    return zero_count


if __name__ == "__main__":
    # Solve the actual puzzle
    result = solve_safe_combination('input.txt')
    print(f"The actual password is: {result}")

File: 2025/01/solution.rs
use std::fs;

fn solve_safe_combination(filename: &str) -> i32 {
    let contents = fs::read_to_string(filename)
        .expect("Failed to read input file");

    let mut position: i32 = 50;
    let mut zero_count = 0;

    for line in contents.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let direction = line.chars().next().unwrap();
        let distance: i32 = line[1..].parse().unwrap();

        if direction == 'L' {
            position = (position - distance).rem_euclid(100);
        } else {
            position = (position + distance).rem_euclid(100);
        }

        if position == 0 {
            zero_count += 1;
        }
    }

    zero_count
}

fn main() {
    // Solve the actual puzzle
    let result = solve_safe_combination("input.txt");
    println!("\nThe actual password is: {}", result);
}
